stan/math/prim/scal/fun/std_binormal_integral.hpp

```diff
@@ -0,0 +1,208 @@
+#include <boost/math/quadrature/tanh_sinh.hpp>
+#include <cmath>
+#include <iostream>
+
+namespace stan {
+namespace math {
+
+/**
+ * The CDF of the standard bivariate normal (binormal) for the specified
+ * variable 1, z1, and variable 2, z2, given the specified correlation, rho.
+ *
+ * The function calculates the bivariate integral with Owen's algorithm for the
+ * domain fabs(z1) <= 1, fabs(z2) <= 1, and fabs(rho) < 0.7, and elsewhere uses
+ * the result std_binormal_integral(z1, z2, rho):
+ *
+ * \f[
+ *   \Phi(z_1)\Phi(z_2) +
+ *   \frac{1}{2\pi}\int_0^{\sin^{-1}(\rho)}\exp
+ *   (\tfrac{-z_1^2 - z_2^2 + 2z_1 z_2 \sin(\theta)}{2\cos(\theta)^2}) d\theta
+ * \f]
+ *
+ * where the integral is done using boost's tanh-sinh quadrature. References
+ * for the integral can be found in:
+ *
+ * Genz, Alan. "Numerical computation of rectangular bivariate
+ * and trivariate normal and t probabilities."
+ * Statistics and Computing 14.3 (2004): 251-260.
+ *
+ * @param z1 Random variable 1
+ * @param z2 Random variable 2
+ * @param rho correlation parameter
+ * for the standard bivariate normal distribution.
+ * @return The probability P(Y1 <= y[1], Y2 <= y[2] | rho).
+ * @throw std::domain_error if the rho is not between -1 and 1 or nan.
+ */
+inline double std_binormal_integral(const double z1, const double z2,
+                                    const double rho) {
+  static const char* function = "std_binormal_integral";
+  using std::asin;
+  using std::exp;
+  using std::fabs;
+  using std::min;
+  using std::sqrt;
+
+  check_not_nan(function, "Random variable 1", z1);
+  check_not_nan(function, "Random variable 2", z2);
+  check_bounded(function, "Correlation coefficient", rho, -1, 1);
+  if (z1 > 10 && z2 > 10)
+    return 1;
+  if (z1 < -37.5 || z2 < -37.5)
+    return 0;
+  if (rho == 0)
+    return Phi(z1) * Phi(z2);
+  if (z1 == rho * z2)
+    return 0.5 / pi() * exp(-0.5 * z2 * z2) * asin(rho) + Phi(z1) * Phi(z2);
+  if (z1 * rho == z2)
+    return 0.5 / pi() * exp(-0.5 * z1 * z1) * asin(rho) + Phi(z1) * Phi(z2);
+  if (z2 > 40)
+    return Phi(z1);
+  if (z1 > 40)
+    return Phi(z2);
+  if (fabs(rho) < 0.7 && fabs(z1) <= 1 && fabs(z2) <= 1) {
+    double denom = sqrt((1 + rho) * (1 - rho));
+    double a1 = (z2 / z1 - rho) / denom;
+    double a2 = (z1 / z2 - rho) / denom;
+    double product = z1 * z2;
+    double delta = product < 0 || (product == 0 && (z1 + z2) < 0);
+    return 0.5 * (Phi(z1) + Phi(z2) - delta) - owens_t(z1, a1)
+           - owens_t(z2, a2);
+  }
+  if (fabs(rho) < 1) {
+    int s = rho > 0 ? 1 : -1;
+    double h = -z1;
+    double k = -z2;
+
+    auto f = [&h, &k](const double x) {
+      using std::cos;
+      using std::sin;
+      using stan::math::square;
+      return exp(0.5 * (-h * h - k * k + h * k * 2 * sin(x)) / square(cos(x)));
+    };
+    double L1, error;
+    boost::math::quadrature::tanh_sinh<double> integrator;
+    double accum
+        = s == 1 ? integrator.integrate(f, 0, asin(rho),
```

Comment:
[optional]
I think it'd be better to not define `s` and replace

* `s == 1` with `rho > 0`

* `s * 0.5 / pi()` with `rho > 0 ? 0.5 * pi() : -0.5 * pi()`